### ISSUE TITLE
Handling multiple routes per message in v6

### ISSUE THREAD

**Issue by timbussmann** [Particular/NServiceBus#3909]
It is possible to configure routing in a way that it yields multiple valid routes for an outgoing message, e.g. when using message hierarchies like this:

    public class MyMessage : ICommand
    {
    }

    public class MySubMessage : MyMessage
    {
    }

with the following routing config:

     routingSettings.RouteToEndpoint(typeof(MyMessage), "v6-1");
     routingSettings.RouteToEndpoint(typeof(MySubMessage), "v6-2");

a send a message to both addresses when calling `endpoint.Send(new MySubMessage());`

V5 seems to send a message only to the "closest matching" rule, e.g. the following endpoint mapping:

    <MessageEndpointMappings>
      <add Assembly="v5-sample"
           Type="MyMessage"
           Endpoint="v5-1" />
      <add Assembly="v5-sample"
           Type="MySubMessage"
           Endpoint="v5-2" />
    </MessageEndpointMappings>

will only send to `v5-2`.

(when using the message endpoint mappings in the v6 sample without the code first API, it will send the message also to both receivers, it's also the same behavior if you mix endpoint mappings and code first API).

Especially regarding the fact that it will be quite easy to have multiple routes matching a specific type when we provide assembly based code first routing APIs, this seems like something we should align with v5? @Particular/nservicebus-maintainers @SzymonPobiega

**Comment by timbussmann** [comment]
btw. I think the same scenario can happen with `RegisterPublisher` which ends with the endpoint sending subscription messages to multiple addresses.

At first glance it would make sense to catch configurations like this at configuration time and throw an exception because of ambiguous routes (shouldn't be to difficult for static routes, dynamic routes can only be handled at runtime)

**Comment by DavidBoike** [comment]
Also here, I think we should align with V5. Especially given the mixture of assembly and type driven rules.

In the wild, this would be caused by starting out with routing an entire assembly to a given endpoint (endpoint would have 1:1 mapping with logical service, in this case) while things were simple, and then realizing that that was cool except for this one message type that needs to be scaled out, so it needs to route somewhere else.

**Comment by danielmarbach** [comment]
How this will be fixed depends on the outcome of https://github.com/Particular/NServiceBus/issues/3911

**Comment by danielmarbach** [comment]
Marked as discussion

**Comment by timbussmann** [comment]
For static routes an obvious approach would be to just overwrite existing routes for a type, which results in a "last one wins" behavior aligning with v5 (and easy to understand).

But how do we deal with dynamic routes and combinations of dynamic + static / dynamic + async dynamic routes? Does it make sense to throw an exception when finding more than one route or should we continue to send to multiple destinations in that case? @DavidBoike @bording

**Comment by SzymonPobiega** [comment]
@timbussmann I think overriding is not the best approach as this invites black magic in form of "route all these messages here but those specific two there and this one over here". How about a simple check that for a command there cane be only one **logical** route selected? Same for the subscribe. There should be no case when there is more than one **logical** subscriber.

**Comment by timbussmann** [comment]
> I think overriding is not the best approach as this invites black magic in form of "route all these messages here but those specific two there and this one over here"

@SzymonPobiega but that's how v5 currently works. I think it may make some sense when combining assembly based message routes but you want to have a specific message routes somewhere else. Otherwise you'd have to register each message type but in general I could still imagine NSB throwing an exception on ambiguous static routes.

> How about a simple check that for a command there cane be only one logical route selected

what kind of check are you thinking about? Do you want to execute dynamic rules at startup or do you just mean some runtime exception when sending a command?

**Comment by SzymonPobiega** [comment]
> what kind of check are you thinking about? Do you want to execute dynamic rules at startup or do you just mean some runtime exception when sending a command?

Yeah I thought about a runtime check when we discover that a route is ambiguous. I feel a bit uncomfortable with the concept of prioritizing routes based on specificity.

**Comment by timbussmann** [comment]
> Yeah I thought about a runtime check when we discover that a route is ambiguous

@SzymonPobiega I'm totally fine with that approach with all dynamically resolved routes or combinations of static + dynamic routes. Since that's all new features I don't see any issues with that. For the static routes, I propose to continue with the overriding mechanism and raise a new issue to discuss this behavior (another approach may be explicit approval by the user to override static routes to avoid accidental overrides)?

**Comment by SzymonPobiega** [comment]
@timbussmann I see the value in being compatible with V5. Maybe the overriding should happen in the MEM handling in V6? I mean, if you just upgrade to V6 and continue using MEMs, you should not experience any difference in behavior but when you convert the routes to the code API you'll have to make them unambiguous.

**Comment by timbussmann** [comment]
> if you just upgrade to V6 and continue using MEMs, you should not experience any difference in behavior but when you convert the routes to the code API you'll have to make them unambiguous.

from an API perspective this would mean that we would have to provide some kind of APIs which allow/disallow overriding anyway (through a bool flag?) anyway, so why not let the user decide (since this wouldn't add any additional to the code complexity).

e.g. something like:
* `config.RouteTo(messageAssembly, "somewhere")` (contains `Message` class)
* `config.RouteTo(typeof(Message), "somwhere else", true)` -> overrides assembly based config
* `config.RouteTo(typeof(Message), "somewhere else")` -> throws an exception because of ambiguous routes
(downside would be the default parameter)

**Comment by SzymonPobiega** [comment]
@timbussmann I was thinking about handling the overriding in the `RoutingFeature` in `ImportMessageEndpointMappings`. This way we would not have to change the API.

**Comment by bording** [comment]
Talked with @timbussmann about this, and the behavior I think makes the most sense here is to default to throwing, and then provide a flag on the API to make the overriding behavior explicit. This should let us handle all the cases of mimicking V5 behavior when it makes sense, and still have an API that is intentional in how it works.

**Comment by SzymonPobiega** [comment]
I don't like `true` because I can't say what does it mean just looking at the code. How about:

```
config.RouteTo(typeof(Message), "somwhere else", RouteType.Override)
```
or
```
config.RouteTo(typeof(Message), "somwhere else", Route.Priority(1)) //The bigger the more important
```

Because I still believe route overriding is a bad pattern, I think the best approach would be to handle route overriding for MEMs only to stay compatible and save users from surprises and expose a clean API in V6.

**Comment by timbussmann** [comment]
@SzymonPobiega the idea was to use an additional overload with the bool flag, so regular users don't have to use it, but if you want to allow overrides, you can explicitly state it. (again, I think we need that API anyway, let's discuss on the PR)

**Comment by DavidBoike** [comment]
Route overriding is not a bad pattern. It's completely normal.

You start off with a service represented entirely by one endpoint, and all that service's messages are routed to it. This works fine until you have one message type that has a different SLA, or needs to be scaled independently. So you reroute that one message to a different endpoint. All messages from this assembly go here, except for this one message that's special and goes over here instead.

What's so bad about specific type routes overriding namespace/assembly routes, and namespace routes overriding assembly routes, and any ambiguity within one of those categories throwing an exception?

I see any sort of 3rd parameter to be unnecessary and over-engineered.

**Comment by timbussmann** [comment]
The new behavior with #3987 is as follows:
* Message Endpoint Mappings override each other as in v5. But there is only one route for a command at any time.
* Using the RouteTo routing API will throw an exception when registering an ambiguous route.
* You can no longer have multiple dynamically evaluated routes but only one, which is only executed when no static route was found.
* The APIs only allow to return a single route for a command
* Subscriptions can still receive multiple routes.

@SzymonPobiega is investigating further changes to dynamic routes, but this will not directly affect the behavior on ambiguous routes.

Should we make further changes to this current behavior (e.g. allow overrides on the code first routing API as @DavidBoike suggests)?

**Comment by timbussmann** [comment]
the following changes were made after #3987:

* The code first API has a route priority, allowing to override routes within certain boundraries:
  * More specific routes override less specific routes. The order is: assembly < namespace < type, where type is the highest priority.
  * Registering two routes for the same type with the same priority still yields an ambiguous route exception.

@SzymonPobiega @DavidBoike @bording anything else we changed around this? I think we can close this issue as there can no longer be multiple routes for a command even from a technical perspective.

**Comment by SzymonPobiega** [comment]
:+1: for closing.
